### DISCUSSION TITLE
How to (temporarily) override / redirect the console.log method?

### DISCUSSION THREAD

**Discussion by boltex** [microsoft/vscode-discussions#1224] [Extension Development QnA]
I'm writing a vscode extension. I want to temporarily override the console.log behavior, so I did this :

            this.old = console.log;
            console.log = this.write.bind(this);

('write' is a method of a class in my extensions, as you see, I kept the real console.log in an 'old' variable to set it back to normal shortly afterward.)

Problem is that this does not seem to actually reassign the "log" method of the console.

Is the console object instance protected in extensionHosts process of vscode or something?

**Comment by isidorn** [comment]
I have no clue to be honest. Do you still see face this challenge?
Why do you actually want to do this?

**Reply by boltex** [reply]
Because my extension (leojs) allows the user to run 'live scripts' in vscode, (see https://youtu.be/M_mKXSbVGdE?t=138 It's similar to 'eval' but I use function() which is safer) and one of my features is to allow the 'console.log' outputs to be redirected to an output pane. (which would only appear otherwise in the developper's console) 

so that why I wanted to temporarily redirect the console.log.

**Reply by isidorn** [reply]
Maybe @hediet has an idea.

**Comment by ArturoDent** [comment]
There are a bunch of search results for overriding console.log, see e.g., https://ourcodeworld.com/articles/read/104/how-to-override-the-console-methods-in-javascript

Perhaps something else will work:

https://stackoverflow.com/questions/35548092/override-console-log-in-javascript

https://stackoverflow.com/questions/7042611/override-console-log-for-production

**Comment by boltex** [comment]
@isidorn Respectfully, Don't know why you marked this as 'answered' with that suggestion of just overriding it as usual, because In the extension host , this is not possible, thus the reason of this discussion in the first place...  The correct answer would be that this is just not possible when running as an extension in vscode. (as far as we know so far)

**Reply by isidorn** [reply]
Thanks. Unmarking the answer then.

**Comment by vaclavHala** [comment]
@boltex  You said you don't use `eval` so the following may not be a feasible solution for you, but maybe you will be able to modify it somehow to fit your needs:

```ts
const out = vscode.window.createOutputChannel("Custom Console")

function withCustomConsole(fn: any) { // you probably want better typing than `any` here for real code
    // This bit of magic runs fn in context where the function argument
    // `console` shadows the global `console`, you probably also need to
    // define or delegate all the rest of console methods like `error` etc.
    const customConsole = {
        log:(msg: any) => out.appendLine(msg)
    }
    return eval(`
        (function(console) {
            return (${fn.toString()})();
        })
    `)(customConsole);
}

// This is arbitrary code provided to you by the user you wish to evaluate.
// Notice console at any depth inside the user script gets intercepted.
// Be aware though this trick relies on userScript.toString() returning actual text
// of the function (which should always be the case for user defined functions).
// Based on your usage maybe you already start with the user function as string
// rather than actual js function object so the conversion to string is not needed at all.
const userScript = () => {
    console.log(1+1)
    function fun(){
        console.log(1+2)
    }
    fun()
}

withCustomConsole(userScript)
// prints into 'Custom Console' output ```
// 2
// 3
// ```


```

**Reply by vaclavHala** [reply]
FYI turns out the same trick works when using the `new Function` syntax also. You can replace the `eval` part above like this:
```js
return new Function("console", `return (${fn.toString()})();`)(customConsole);
```

**Comment by theReynald** [comment]
Sounds like the conversation has come to an end. Closing. If more discussion is needed, please reopen.
